keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+  - expression: "object.spec.replicas < 100"
+```
+
+#### Use Case: Shared Configuration
+
+User wishes to define a configuration for a list of banned words that may not be
+used in any of a wide range of identifiers in the cluster (resource names,
+container names, ...).
+
+- Configuration CRD is defined to hold the list of banned words.
+- Multiple policies are defined for different resources. The policies all
+  reference the same configuration CRD. 
+- Policy must be able to specify it's own matching rules since the configuration
+  applies cluster wide.
+- A single custom resource is defined with the list of banned words (but has not
+  matching rules of its own).
+
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: BannedWords
+metadata:
+  name: "banned-words.example.com"
+spec:
+  bannedWords:
+  - glitter
+  - rainbow
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy1.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   ["*"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+      scope:       "*"
+  validations:
+  - rule: "!object.name in config.bannedWords"
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy2.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+      scope:       "*"
+  validations:
+  - rule: "!object.spec.containers.any(c, c.name in config.bannedNames)"
+  - rule: "!object.spec.initContainers.any(c, c.name in config.bannedNames)"
+```
+
+
+Similar Use Case: A cluster administrator wishes to use a single policy
+configuration to manage a network policy that must be enforced across multiple
+Kubernetes kinds that contain relevant networking fields. It is possible to
+implement by having multiple `ValidatingAdmissionPolicy` resources that all
+reference the same `spec.config` CRD but that each enforce the policy for a
+different Kubernetes network kind.
+
+#### Use Case: Principle of least privilege policy
+
+A cluster administrator would like disallow the use of a list of reserved labels
+by default, but allow use of the labels in specific namespaces so long as the
+label values are valid. 
+
+- Define the following policy configurations:
+  - policy configuration that matches all namespaces with the
+    "may-use-reserved-labels" label and checks that the values are valid
+  - "default" policy configuration that matches all namespaces without the
+    "may-use-reserved-labels" label and disallows any reserved labels.
+  - An additional policy can be used validate that only authorized roles set the
+    "may-use-reserved-labels".
+
+See "Default configuration
```

Comment:
given mutually exclusive labels, you don't actually need default configuration, right?

Reply:
Correct, if we did do default configurations, it wouldn't be unlocking functionality, it would be purely for ergonomics. Also, I haven't found any sane designs that would make this possible. I'm going to drop it.